rustls/src/hash_hs.rs

```diff
@@ -197,6 +207,42 @@ test_for_each_provider! {
         assert_eq!(h[3], 0x5c);
     }
 
+    #[test]
+    fn hashes_message_types() {
+        // handshake protocol encoding of 0x0e 00 00 00
+        let server_hello_done_message = Message {
+            version: ProtocolVersion::TLSv1_2,
+            payload: MessagePayload::handshake(HandshakeMessagePayload {
+                        typ: HandshakeType::ServerHelloDone,
+                        payload: HandshakePayload::ServerHelloDone,
+                        }) };
+
+        let app_data_ignored = Message {
+            version: ProtocolVersion::TLSv1_3,
+            payload: MessagePayload::ApplicationData(Payload::Borrowed(b"hello")),
+        };
+
+        let end_of_early_data_flight = Message {
+            version: ProtocolVersion::TLSv1_3,
+            payload: MessagePayload::handshake_flight(b"\x05\x00\x00\x00".to_vec())
+        };
+
+        let mut hhb = HandshakeHashBuffer::new();
+        hhb.add_message(&server_hello_done_message);
+        hhb.add_message(&app_data_ignored);
+        hhb.add_message(&end_of_early_data_flight);
+        assert_eq!(hhb.start_hash(&SHA256).current_hash().as_ref(),
+                   SHA256.hash(b"\x0e\x00\x00\x00\x05\x00\x00\x00").as_ref());
+
+        let mut hh = HandshakeHashBuffer::new().start_hash(&SHA256);
+        hh.add_message(&server_hello_done_message);
+        hh.add_message(&app_data_ignored);
+        hh.add_message(&end_of_early_data_flight);
+        assert_eq!(hh.current_hash().as_ref(),
+                   SHA256.hash(b"\x0e\x00\x00\x00\x05\x00\x00\x00").as_ref());
```

Comment:
I think the difference between these two test cases is in the coverage they produce, but it's not obvious from a first glance. Worth a comment that calls out the difference in approach for A vs B?